File: algorithm/18. 4Sum/18.py
class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        size=len(nums)
        result=[]
        
        if(size<4):
            return result
        nums.sort()
        for i in xrange(size-3):
            if i>0 and nums[i]==nums[i-1]:
                continue
            for j in range(i+1,size-2):
                if j > i+1  and nums[j]==nums[j-1]:
                    continue
                k=j+1
                l=size-1
                while k<l:
                    if nums[i]+nums[j]+nums[k]+nums[l]==target:
                        result.append([nums[i], nums[j], nums[k],nums[l]])
                        k+=1
                        l-=1
                        while k < l and nums[k] == nums[k -1 ]:
                            k += 1
                        while k < l and nums[l] == nums[l + 1]:
                            l -= 1
                    elif nums[i]+nums[j]+nums[k]+nums[l]<target:
                        k += 1
                    else:
                        l -= 1
        return result
            

File: algorithm/18. 4Sum/18.java
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        Arrays.sort(nums);
        List<List<Integer>> res = new LinkedList<>(); 
        for(int i = 0; i < nums.length; i++)
        {
            if(i > 0 && nums[i] == nums[i - 1]) continue;
            for(int j = i + 1; j < nums.length; j++)
            {
                if(j > i + 1 && nums[j] == nums[j - 1]) continue;
                int left = j + 1, right = nums.length - 1;
                while(left < right)
                {
                    int tem = nums[i] + nums[j] + nums[left] + nums[right];
                    if(tem == target)
                    {
                        res.add(Arrays.asList(nums[i], nums[j], nums[left], nums[right]));
                        while(++left < right && nums[left] == nums[left - 1]);
                        while(left < --right && nums[right] == nums[right + 1]);
                    }
                    else if(tem < target)
                        left++;
                    else
                        right--;
                }
            }
        }
        return res;
    }
}
